### ISSUE TITLE
On the efficiency of code

### ISSUE THREAD

**Issue by time-beyond** [BroadbandForum/obuspa#16]
hello:
I have a question. Why do you call the same interface(DATABASE_SetParameterValue) for add and set operations in the database layer? You should know that replace affects efficiency. And when add, is it reasonable to skip the parameter of type kDMNodeType_DBParam_ReadOnly directly?
                                                                                                                                            Looking forward to your reply

**Comment by time-beyond** [comment]
![image](https://user-images.githubusercontent.com/54758785/110594927-5f6ee200-81b8-11eb-83ab-974015e15cdc.png)
![Uploading image.png…]()

**Comment by r-holme** [comment]
On your first point: That's an interesting efficiency improvement. I think we would have to benchmark the overall speed increase in the context of the full time taken to process a USP Add message.
On your second point: Please can you expand on what you mean by "And when add, is it reasonable to skip the parameter of type kDMNodeType_DBParam_ReadOnly directly?"

**Comment by time-beyond** [comment]
For the first problem, since our system needs to reach 2ms level, when I call DATABASE_GetParameterValue，i found that the execution time of SQL reached 50ms. Through analysis, it was found that it was caused by “insert or replace‘’. Do you have no strict requirements on the execution time of the DATABASE_GetParameterValue？
Second problem, you can look at the function AddChildParamsDefaultValues，
_**“ // Nothing to do for non database writable parameters
            default:      
            case kDMNodeType_DBParam_ReadOnly:
	   case kDMNodeType_VendorParam_ReadOnly:
            case kDMNodeType_VendorParam_ReadWrite:
            case kDMNodeType_Param_ConstantValue:
            case kDMNodeType_Param_NumEntries:
            case kDMNodeType_SyncOperation:
            case kDMNodeType_AsyncOperation:
            case kDMNodeType_Event:
            case kDMNodeType_Object_MultiInstance:  // A child Multi-instance object needs an instance creating before we can populate it**_
                break;”
why kDMNodeType_DBParam_ReadOnly also break ？

**Comment by time-beyond** [comment]
I think kDMNodeType_DBParam_ReadOnly should be like kDMNodeType_DBParam_ReadWrite，it should  insert into database

**Comment by time-beyond** [comment]
Thanks for your help.
But for the first question, you don't think that when you perform a set operation, the code uses 'replace', which is far less rapid than 'update'. Because our system requires high efficiency.
Did you think about this at the time？
                                                                                                                                     We look forward to your reply

**Comment by time-beyond** [comment]
replace：
1. If this row of data is found in the table (judged by primary key or unique index), delete this row of data first, and then insert new data.
2. Otherwise, insert new data directly.

**Comment by holme-r** [comment]
We've raised a ticket to work on the add object efficiency improvement you propose.

**Comment by holme-r** [comment]
We performed some profiling tests and came to the following conclusions:
* USP Add handling is not improved by using SQL 'insert' instead of 'insert or replace'. This is because the SQL statement is only run if OBUSPA knows that the object's parameter is not in the database already. Effectively for an Add, the SQL only ever performs an 'insert'.
* USP Set handling could be improved by using SQL 'update' instead of 'insert or replace'. However there are many cases where parameters may not necessarily be present in the database though the parameter is present in the instantiated data model. For example if not explicitly set in a factory reset database, the parameter is not stored in the database - a USP Get returns the default value instead. Since it is not possible to know whether the parameter is present in the database when performing a USP Set, OBUSPA doesn't know whether to attempt a SQL 'insert' or 'update'. It is faster to perform an 'insert or replace' than an 'update' which fails, followed by an 'insert'. If we used 'update', followed by 'insert' on failure, we would be optimizing for sets on non defaulted parameters at the expense of sets on defaulted parameters. USP Sets are mainly for tweaking parameters after they have been added, and therefore we expect would be used with much less frequency than USP Adds. So it makes sense to optimize for USP Adds.  Overall it is not easy to change OBUSPA to use SQL 'update' instead of 'insert or replace' for USP Set, nor is it desirable in all cases. So we do not intend to make this change.
